core/src/main/java/org/apache/carbondata/core/datastore/page/UnsafeVarLengthColumnPage.java

```diff
@@ -0,0 +1,124 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.datastore.page;
+
+import java.math.BigDecimal;
+
+import org.apache.carbondata.core.memory.CarbonUnsafe;
+import org.apache.carbondata.core.memory.MemoryBlock;
+import org.apache.carbondata.core.memory.MemoryException;
+import org.apache.carbondata.core.memory.UnsafeMemoryManager;
+import org.apache.carbondata.core.metadata.datatype.DataType;
+import org.apache.carbondata.core.util.DataTypeUtil;
+
+// This extension uses unsafe memory to store page data, for variable length data type (string,
+// decimal)
+public class UnsafeVarLengthColumnPage extends VarLengthColumnPageBase {
+
+  // memory allocated by Unsafe
+  private MemoryBlock memoryBlock;
+
+  // base address of memoryBlock
+  private Object baseAddress;
+
+  // base offset of memoryBlock
+  private long baseOffset;
+
+  // size of the allocated memory, in bytes
+  private int capacity;
+
+  // default size for each row, grows as needed
+  private static final int DEFAULT_ROW_SIZE = 8;
+
+  private static final double FACTOR = 1.25;
+
+  UnsafeVarLengthColumnPage(DataType dataType, int pageSize) throws MemoryException {
+    super(dataType, pageSize);
+    capacity = pageSize * DEFAULT_ROW_SIZE;
+    memoryBlock = UnsafeMemoryManager.allocateMemoryWithRetry((long)(capacity * FACTOR));
+    baseAddress = memoryBlock.getBaseObject();
+    baseOffset = memoryBlock.getBaseOffset();
+  }
+
+  @Override
+  public void freeMemory() {
+    if (memoryBlock != null) {
+      UnsafeMemoryManager.INSTANCE.freeMemory(memoryBlock);
+      memoryBlock = null;
+      baseAddress = null;
+      baseOffset = 0;
+    }
+  }
+
+  private void ensureMemory(int requestSize) throws MemoryException {
+    if (totalLength + requestSize > capacity) {
+      memoryBlock = UnsafeMemoryManager.reallocateMemoryWithRetry(
+          memoryBlock, (long)((totalLength + requestSize) * FACTOR));
+      baseAddress = memoryBlock.getBaseObject();
+      baseOffset = memoryBlock.getBaseOffset();
+    }
+  }
+
+  @Override
+  public void putBytesAtRow(int rowId, byte[] bytes) {
+    try {
+      ensureMemory(bytes.length);
+    } catch (MemoryException e) {
+      throw new RuntimeException(e);
+    }
+
+    CarbonUnsafe.unsafe.copyMemory(bytes, CarbonUnsafe.BYTE_ARRAY_OFFSET,
+        baseAddress, baseOffset + rowOffset[rowId], bytes.length);
+  }
+
+  @Override
+  public void putBytes(int rowId, byte[] bytes, int offset, int length) {
+    CarbonUnsafe.unsafe.copyMemory(bytes, CarbonUnsafe.BYTE_ARRAY_OFFSET + offset,
+        baseAddress, baseOffset + rowOffset[rowId], length);
+  }
+
```

Comment:
Even method `public void putBytes(int rowId, byte[] bytes)` also need to be override here.